File: IT3011_DSA/code/chap1/tim_tong_day_con_nho_nhat/C++/main.cpp
#include <iostream>
#include <climits>
#include <chrono>

using namespace std;

void brute_force(int arr[], int n)
{

    int maxSum = INT_MIN;
    for (int i = 0; i < n; i++)
    {
        for (int j = i; j < n; j++)
        {
            int sum = 0;
            for (int k = i; k <= j; k++)
            {
                sum += arr[k];
            }
            if (sum > maxSum)
                maxSum = sum;
        }
    }
    cout << maxSum << endl;
}

void improved_brute_force(int arr[], int n)
{
    int maxSum = arr[0];
    for (int i = 0; i < n; i++)
    {
        int sum = 0;
        for (int j = i; j < n; j++)
        {
            sum += arr[j];
            if (sum > maxSum)
                maxSum = sum;
        }
    }
    cout << maxSum << endl;
}

int max_left(int arr[], int low, int high){
    int max_sum = INT_MIN;
    int sum = 0;
    for (int k=high; k>=low; k--){
        sum += arr[k];
        max_sum = max(sum, max_sum);
    }
    return max_sum;
}


int max_right(int arr[], int low, int high){
    int max_sum = INT_MIN;
    int sum = 0;
    for (int k=low; k<=high; k++){
        sum += arr[k];
        max_sum = max(sum, max_sum);
    }
    return max_sum;
}

int divide_and_conquer(int arr[], int low, int high){
    if (low == high) return arr[low];
    else {
        int mid = (low + high) / 2;
        int wL = divide_and_conquer(arr, low, mid);
        int wR = divide_and_conquer(arr, mid + 1, high);
        int wM = max_left(arr, low, mid) + max_right(arr, mid+1, high);
        return max(max(wL, wR), wM);
    }
}

int dynamic_programming(int arr[], int n){
    int smax = arr[0];
    int ei = arr[0];
    for (int i=1; i<n;i++){
        // int u = ei + arr[i];
        // int v = arr[i];
        // if (u > v) ei = u;
        // else ei = v;
        // if (ei >smax) {
        //     smax = ei;
        //     imax = i;
        // }
        ei = max(arr[i], ei + arr[i]);
        smax = max(smax, ei);
    }
    return smax;
}

int main()
{

    int n;
    cin >> n;
    int arr[n];
    for (int j = 0; j < n; j++)
    {
        cin >> arr[j];
    }
    // cout << caculate_execution_time(brute_force, arr, n);
    // brute_force(arr, n);
    // improved_brute_force(arr, n);
    // cout << divide_and_conquer(arr, 0, n-1);
    cout << dynamic_programming(arr, n);
    return 0;
}

File: IT3011_DSA/code/chap1/tim_tong_day_con_nho_nhat/Python/main.py
import sys


def brute_force(arr, n):
    max_sum = - sys.maxsize - 1
    for i in range(n):
        current_sum = 0
        for j in range(i, n):
            current_sum += arr[j]
            if (current_sum > max_sum):
                max_sum = current_sum

    return max_sum

def max_left(arr, low, high):
    max_sum = - sys.maxsize - 1
    current_sum = 0
    for k in range(high, low - 1, -1):
        current_sum += arr[k]
        max_sum = max(max_sum, current_sum)
    return max_sum

def max_right(arr, low, high):
    max_sum = - sys.maxsize - 1
    current_sum = 0
    for k in range(low, high + 1):
        current_sum += arr[k]
        max_sum = max(max_sum, current_sum)
    return max_sum

def divide_and_conquer(arr, low, high):
    if (low == high):
        return arr[low]
    else:
        mid = (low + high) // 2
        wl = divide_and_conquer(arr, low, mid)
        wR = divide_and_conquer(arr, mid + 1, high)
        wM = max_left(arr, low, mid) + max_right(arr, mid + 1, high)
        return max(wl, wR, wM)

def dynamic_programming(arr, n):
    smax = arr[0]
    ei = arr[0]
    for i in range(1, n):
        ei = max(arr[i], ei + arr[i])
        smax = max(smax, ei)
    return smax

if __name__ == "__main__":
    n = int(input().strip())
    arr = list(map(int, input().split()))
    # print(brute_force(arr, n))
    # print(divide_and_conquer(arr, 0, n-1))
    print(dynamic_programming(arr, n))
